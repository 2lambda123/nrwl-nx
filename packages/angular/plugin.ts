export * from './plugins/nx-plugin';
