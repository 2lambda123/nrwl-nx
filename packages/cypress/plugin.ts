export { createNodes, NxCypressMetadata } from './src/plugins/plugin';
