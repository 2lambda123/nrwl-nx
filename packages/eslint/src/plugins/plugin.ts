import {
  CreateNodes,
  CreateNodesContext,
  TargetConfiguration,
} from '@nx/devkit';
import { basename, dirname, join } from 'path';
import { readTargetDefaultsForTarget } from 'nx/src/project-graph/utils/project-configuration-utils';
import { readdirSync } from 'fs';
import { combineGlobPatterns } from 'nx/src/utils/globs';
import {
  ESLINT_CONFIG_FILENAMES,
  findBaseEslintFile,
  isFlatConfig,
} from '../utils/config-file';

export interface EslintPluginOptions {
  targetName?: string;
  excludedProjects?: string[];
}

export const createNodes: CreateNodes<EslintPluginOptions> = [
  combineGlobPatterns(['**/project.json', '**/package.json']),
  (configFilePath, options, context) => {
    const projectRoot = dirname(configFilePath);

    if (!projectHasEslintConfig(projectRoot, context.workspaceRoot)) {
      return {};
    }

    options = normalizeOptions(options);
    const projectName = basename(projectRoot);

    const rootEslintConfigFile = findBaseEslintFile(context.workspaceRoot);

    return {
      projects: {
        [projectName]: {
          root: projectRoot,
          targets: buildEslintTargets(
            rootEslintConfigFile,
            projectRoot,
            options,
            context
          ),
        },
      },
    };
  },
];

function projectHasEslintConfig(
  projectRoot: string,
  workspaceRoot: string
): boolean {
  const siblingFiles = readdirSync(join(workspaceRoot, projectRoot));

  if (projectRoot === '.') {
    // If there's no src folder, it's not a standalone project
    if (!siblingFiles.includes('src')) {
      return false;
    }
    // If it's standalone but doesn't have eslint config, it's not a lintable
    if (!siblingFiles.some((f) => ESLINT_CONFIG_FILENAMES.includes(f))) {
      return false;
    }
    return true;
  }
  // if it has an eslint config it's lintable
  if (siblingFiles.some((f) => ESLINT_CONFIG_FILENAMES.includes(f))) {
    return true;
  }
  // check whether the root has an eslint config
  return readdirSync(workspaceRoot).some((f) =>
    ESLINT_CONFIG_FILENAMES.includes(f)
  );
}

function buildEslintTargets(
  rootEslintConfigFile: string,
  projectRoot: string,
  options: EslintPluginOptions,
  context: CreateNodesContext
) {
  const targetDefaults = readTargetDefaultsForTarget(
    options.targetName,
    context.nxJsonConfiguration.targetDefaults,
    '@nx/eslint:lint'
  );

  const isRootProject = projectRoot === '.';

  const targets: Record<string, TargetConfiguration> = {};

  const baseTargetConfig: TargetConfiguration = {
    command: `eslint ${isRootProject ? './src' : '.'}`,
    options: {
      cwd: projectRoot,
    },
  };
  if (isFlatConfig(rootEslintConfigFile)) {
    baseTargetConfig.options.env = {
      ESLINT_USE_FLAT_CONFIG: 'true',
    };
  }

  targets[options.targetName] = {
    ...baseTargetConfig,
    cache: targetDefaults?.cache ?? true,
    inputs: targetDefaults?.inputs ?? [
      'default',
      `{workspaceRoot}/${rootEslintConfigFile}`,
      '{workspaceRoot}/tools/eslint-rules/**/*',
      { externalDependencies: ['eslint'] },
    ],
    options: {
      ...baseTargetConfig.options,
    },
  };

  return targets;
}

function normalizeOptions(options: EslintPluginOptions): EslintPluginOptions {
  options ??= {};
  options.targetName ??= 'lint';
  options.excludedProjects ??= [];
  return options;
}
