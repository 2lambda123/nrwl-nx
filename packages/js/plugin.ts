export {
  createDependencies,
  createNodes,
  type JsPluginOptions,
} from './src/plugins/plugin';
