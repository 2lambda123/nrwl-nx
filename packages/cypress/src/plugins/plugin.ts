import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  readJsonFile,
  TargetConfiguration,
  writeJsonFile,
} from '@nx/devkit';
import { dirname, extname, join } from 'path';
import { registerTsProject } from '@nx/js/src/internal';

import { getRootTsConfigPath } from '@nx/js';

import { CypressExecutorOptions } from '../executors/cypress/cypress.impl';
import { readTargetDefaultsForTarget } from 'nx/src/project-graph/utils/project-configuration-utils';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { existsSync, readdirSync } from 'fs';
import { globWithWorkspaceContext } from 'nx/src/utils/workspace-context';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';

export interface CypressPluginOptions {
  ciTargetName?: string;
  targetName?: string;
  componentTestingTargetName?: string;
}

const cachePath = join(projectGraphCacheDirectory, 'cypress.hash');
const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};

const calculatedTargets: Record<
  string,
  Record<string, TargetConfiguration>
> = {};

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration<CypressExecutorOptions>>
> {
  return readJsonFile(cachePath);
}

function writeTargetsToCache(
  targets: Record<
    string,
    Record<string, TargetConfiguration<CypressExecutorOptions>>
  >
) {
  writeJsonFile(cachePath, targets);
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache(calculatedTargets);
  return [];
};

export const createNodes: CreateNodes<CypressPluginOptions> = [
  '**/cypress.config.{js,ts,mjs,mts,cjs,cts}',
  (configFilePath, options, context) => {
    options = normalizeOptions(options);
    const projectRoot = dirname(configFilePath);

    // Do not create a project if package.json and project.json isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') &&
      !siblingFiles.includes('project.json')
    ) {
      return {};
    }

    const hash = calculateHashForCreateNodes(projectRoot, options, context);

    const targets = targetsCache[hash]
      ? targetsCache[hash]
      : buildCypressTargets(configFilePath, projectRoot, options, context);

    calculatedTargets[hash] = targets;

    return {
      projects: {
        [projectRoot]: {
          projectType: 'application',
          targets,
        },
      },
    };
  },
];

function getOutputs(
  projectRoot: string,
  cypressConfig: any,
  testingType: 'e2e' | 'component'
): string[] {
  function getOutput(path: string): string {
    if (path.startsWith('..')) {
      return join('{workspaceRoot}', join(projectRoot, path));
    } else {
      return join('{projectRoot}', path);
    }
  }

  const { screenshotsFolder, videosFolder, e2e, component } = cypressConfig;
  const outputs = ['{options.videosFolder}', '{options.screenshotsFolder}'];

  if (videosFolder) {
    outputs.push(getOutput(videosFolder));
  }

  if (screenshotsFolder) {
    outputs.push(getOutput(screenshotsFolder));
  }

  switch (testingType) {
    case 'e2e': {
      if (e2e.videosFolder) {
        outputs.push(getOutput(e2e.videosFolder));
      }
      if (e2e.screenshotsFolder) {
        outputs.push(getOutput(e2e.screenshotsFolder));
      }
      break;
    }
    case 'component': {
      if (component.videosFolder) {
        outputs.push(getOutput(component.videosFolder));
      }
      if (component.screenshotsFolder) {
        outputs.push(getOutput(component.screenshotsFolder));
      }
      break;
    }
  }

  return outputs;
}
function buildCypressTargets(
  configFilePath: string,
  projectRoot: string,
  options: CypressPluginOptions,
  context: CreateNodesContext
) {
  const cypressConfig = getCypressConfig(configFilePath, context);

  const namedInputs = getNamedInputs(projectRoot, context);

  const baseTargetConfig: TargetConfiguration<CypressExecutorOptions> = {
    executor: '@nx/cypress:cypress',
    options: {
      cypressConfig: configFilePath,
    },
  };

  const targets: Record<
    string,
    TargetConfiguration<CypressExecutorOptions>
  > = {};

  if ('e2e' in cypressConfig) {
    const e2eTargetDefaults = readTargetDefaultsForTarget(
      options.targetName,
      context.nxJsonConfiguration.targetDefaults,
      '@nx/cypress:cypress'
    );

    targets[options.targetName] = {
      ...baseTargetConfig,
      options: {
        ...baseTargetConfig.options,
        testingType: 'e2e',
      },
    };

    if (e2eTargetDefaults?.cache === undefined) {
      targets[options.targetName].cache = true;
    }

    if (e2eTargetDefaults?.inputs === undefined) {
      targets[options.targetName].inputs =
        'production' in namedInputs
          ? ['default', '^production']
          : ['default', '^default'];
    }

    if (e2eTargetDefaults?.outputs === undefined) {
      targets[options.targetName].outputs = getOutputs(
        projectRoot,
        cypressConfig,
        'e2e'
      );
    }

    const cypressEnv = {
      ...cypressConfig.env,
      ...cypressConfig.e2e?.env,
    };

    const devServerTargets: Record<string, string> =
      cypressEnv?.devServerTargets;

    if (devServerTargets?.default) {
      targets[options.targetName].options.devServerTarget =
        devServerTargets.default;
      delete devServerTargets.default;
    }

    if (Object.keys(devServerTargets ?? {}).length > 0) {
      targets[options.targetName].configurations ??= {};
      for (const [configuration, devServerTarget] of Object.entries(
        devServerTargets ?? {}
      )) {
        targets[options.targetName].configurations[configuration] = {
          devServerTarget,
        };
      }
    }

    const ciDevServerTarget: string = cypressEnv?.ciDevServerTarget;
    if (ciDevServerTarget) {
      const specPatterns = Array.isArray(cypressConfig.e2e.specPattern)
        ? cypressConfig.e2e.specPattern.map((p) => join(projectRoot, p))
        : [join(projectRoot, cypressConfig.e2e.specPattern)];

      const excludeSpecPatterns: string[] = !cypressConfig.e2e
        .excludeSpecPattern
        ? cypressConfig.e2e.excludeSpecPattern
        : Array.isArray(cypressConfig.e2e.excludeSpecPattern)
        ? cypressConfig.e2e.excludeSpecPattern.map((p) => join(projectRoot, p))
        : [join(projectRoot, cypressConfig.e2e.excludeSpecPattern)];
      const specFiles = globWithWorkspaceContext(
        context.workspaceRoot,
        specPatterns,
        excludeSpecPatterns
      );

      const dependsOn: TargetConfiguration['dependsOn'] = [];
      const outputs = getOutputs(projectRoot, cypressConfig, 'e2e');
      const inputs =
        'production' in namedInputs
          ? ['default', '^production']
          : ['default', '^default'];
      for (const file of specFiles) {
        const targetName = options.ciTargetName + '--' + file;
        targets[targetName] = {
          ...targets[options.targetName],
          outputs,
          inputs,
          cache: true,
          configurations: undefined,
          options: {
            ...targets[options.targetName].options,
            devServerTarget: ciDevServerTarget,
            spec: file,
          },
        };
        dependsOn.push({
          target: targetName,
          projects: 'self',
          params: 'forward',
        });
      }
      targets[options.ciTargetName] ??= {};

      targets[options.ciTargetName] = {
        executor: 'nx:noop',
        cache: true,
        inputs,
        outputs,
        dependsOn,
      };
    }
  }

  if ('component' in cypressConfig) {
    const componentTestingTargetDefaults = readTargetDefaultsForTarget(
      options.componentTestingTargetName,
      context.nxJsonConfiguration.targetDefaults,
      '@nx/cypress:cypress'
    );

    // This will not override the e2e target if it is the same
    targets[options.componentTestingTargetName] ??= {
      ...baseTargetConfig,
      options: {
        ...baseTargetConfig.options,
        testingType: 'component',
      },
    };

    if (componentTestingTargetDefaults?.cache === undefined) {
      targets[options.componentTestingTargetName].cache = true;
    }

    if (componentTestingTargetDefaults?.inputs === undefined) {
      targets[options.componentTestingTargetName].inputs =
        'production' in namedInputs
          ? ['default', '^production']
          : ['default', '^default'];
    }

    if (componentTestingTargetDefaults?.outputs === undefined) {
      targets[options.componentTestingTargetName].outputs = getOutputs(
        projectRoot,
        cypressConfig,
        'component'
      );
    }
  }

  return targets;
}

function getCypressConfig(
  configFilePath: string,
  context: CreateNodesContext
): any {
  const resolvedPath = join(context.workspaceRoot, configFilePath);

  let module: any;
  if (['.ts', '.mts', '.cts'].includes(extname(configFilePath))) {
    const tsConfigPath = getRootTsConfigPath();

    if (tsConfigPath) {
      const unregisterTsProject = registerTsProject(tsConfigPath);
      try {
        module = require(resolvedPath);
      } finally {
        unregisterTsProject();
      }
    } else {
      module = require(resolvedPath);
    }
  } else {
    module = require(resolvedPath);
  }
  return module.default ?? module;
}

function normalizeOptions(options: CypressPluginOptions): CypressPluginOptions {
  options ??= {};
  options.targetName ??= 'e2e';
  options.componentTestingTargetName ??= 'component-test';
  options.ciTargetName ??= 'e2e-ci';
  return options;
}
