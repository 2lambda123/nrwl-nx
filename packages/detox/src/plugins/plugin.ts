import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  detectPackageManager,
  NxJsonConfiguration,
  readJsonFile,
  TargetConfiguration,
  writeJsonFile,
} from '@nx/devkit';
import { dirname, join } from 'path';
import { getLockFileName } from '@nx/js';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { existsSync } from 'fs';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { workspaceDataDirectory } from 'nx/src/utils/cache-directory';
import { projectFoundInRootPath } from '@nx/devkit/src/utils/project-found-in-root-path';

export interface DetoxPluginOptions {
  buildTargetName?: string;
  startTargetName?: string;
  testTargetName?: string;
}

const cachePath = join(workspaceDataDirectory, 'detox.hash');
const targetsCache = readTargetsCache();

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration<DetoxPluginOptions>>
> {
  return existsSync(cachePath) ? readJsonFile(cachePath) : {};
}

function writeTargetsToCache() {
  writeJsonFile(cachePath, targetsCache);
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache();
  return [];
};

export const createNodes: CreateNodes<DetoxPluginOptions> = [
  '**/{detox.config,.detoxrc}.{json,js}',
  async (configFilePath, options, context) => {
    options = normalizeOptions(options);
    const projectRoot = dirname(configFilePath);

    // Configurations will be generated only if project exists at projectRoot
    if (!projectFoundInRootPath(projectRoot, context)) {
      return {};
    }

    const hash = await calculateHashForCreateNodes(
      projectRoot,
      options,
      context,
      [getLockFileName(detectPackageManager(context.workspaceRoot))]
    );

    targetsCache[hash] ??= buildDetoxTargets(projectRoot, options, context);

    return {
      projects: {
        [projectRoot]: {
          targets: targetsCache[hash],
        },
      },
    };
  },
];

function buildDetoxTargets(
  projectRoot: string,
  options: DetoxPluginOptions,
  context: CreateNodesContext
) {
  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration> = {
    [options.buildTargetName]: {
      command: `detox build`,
      options: { cwd: projectRoot },
      cache: true,
      inputs: getInputs(namedInputs),
    },
    [options.startTargetName]: {
      command: `detox start`,
      options: { cwd: projectRoot },
    },
    [options.testTargetName]: {
      command: `detox test`,
      options: { cwd: projectRoot },
      cache: true,
      inputs: getInputs(namedInputs),
    },
  };

  return targets;
}

function getInputs(
  namedInputs: NxJsonConfiguration['namedInputs']
): TargetConfiguration['inputs'] {
  return [
    ...('production' in namedInputs
      ? ['default', '^production']
      : ['default', '^default']),
    {
      externalDependencies: ['detox'],
    },
  ];
}

function normalizeOptions(options: DetoxPluginOptions): DetoxPluginOptions {
  options ??= {};
  options.buildTargetName ??= 'build';
  options.startTargetName ??= 'start';
  options.testTargetName ??= 'test';
  return options;
}
