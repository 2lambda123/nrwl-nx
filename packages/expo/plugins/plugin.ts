import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  detectPackageManager,
  NxJsonConfiguration,
  readJsonFile,
  TargetConfiguration,
  workspaceRoot,
  writeJsonFile,
} from '@nx/devkit';
import { dirname, join } from 'path';
import { getLockFileName } from '@nx/js';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { existsSync, readdirSync } from 'fs';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';

export interface ExpoPluginOptions {
  startTargetName?: string;
  serveTargetName?: string;
  runIosTargetName?: string;
  runAndroidTargetName?: string;
  exportTargetName?: string;
  exportWebTargetName?: string;
  prebuildTargetName?: string;
  installTargetName?: string;
  buildTargetName?: string;
  submitTargetName?: string;
}

const cachePath = join(projectGraphCacheDirectory, 'expo.hash');
const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};

const calculatedTargets: Record<
  string,
  Record<string, TargetConfiguration>
> = {};

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration<ExpoPluginOptions>>
> {
  return readJsonFile(cachePath);
}

function writeTargetsToCache(
  targets: Record<
    string,
    Record<string, TargetConfiguration<ExpoPluginOptions>>
  >
) {
  writeJsonFile(cachePath, targets);
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache(calculatedTargets);
  return [];
};

export const createNodes: CreateNodes<ExpoPluginOptions> = [
  '**/app.{json,config.js}',
  (configFilePath, options, context) => {
    options = normalizeOptions(options);
    const projectRoot = dirname(configFilePath);

    // Do not create a project if package.json or project.json or metro.config.js isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') ||
      !siblingFiles.includes('project.json') ||
      !siblingFiles.includes('metro.config.js')
    ) {
      return {};
    }
    const appConfig = getAppConfig(configFilePath, context);
    // if appConfig.expo is not defined
    if (!appConfig.expo) {
      return {};
    }

    const hash = calculateHashForCreateNodes(projectRoot, options, context, [
      getLockFileName(detectPackageManager(context.workspaceRoot)),
    ]);

    const targets = targetsCache[hash]
      ? targetsCache[hash]
      : buildExpoTargets(projectRoot, options, context);

    calculatedTargets[hash] = targets;

    return {
      projects: {
        [projectRoot]: {
          targets,
        },
      },
    };
  },
];

function buildExpoTargets(
  projectRoot: string,
  options: ExpoPluginOptions,
  context: CreateNodesContext
) {
  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration> = {
    [options.startTargetName]: {
      executor: `@nx/expo:start`,
    },
    [options.serveTargetName]: {
      command: `expo start --web`,
      options: { cwd: projectRoot },
    },
    [options.runIosTargetName]: {
      executor: `@nx/expo:run-ios`,
    },
    [options.runAndroidTargetName]: {
      executor: `@nx/expo:run-android`,
    },
    [options.exportTargetName]: {
      command: `expo export`,
      options: { cwd: projectRoot },
      cache: true,
      dependsOn: [`^${options.exportTargetName}`],
      inputs: getInputs(namedInputs),
      outputs: [getOutputs(projectRoot, 'dist')],
    },
    [options.exportWebTargetName]: {
      command: `expo export:web`,
      options: { cwd: projectRoot },
      cache: true,
      dependsOn: [`^${options.exportWebTargetName}`],
      inputs: getInputs(namedInputs),
      outputs: [getOutputs(projectRoot, 'web-build')],
    },
    [options.installTargetName]: {
      command: `expo install`,
      options: { cwd: workspaceRoot }, // install at workspace root
    },
    [options.prebuildTargetName]: {
      executor: `@nx/expo:prebuild`,
    },
    [options.buildTargetName]: {
      executor: `@nx/expo:build`,
      dependsOn: [`^${options.buildTargetName}`],
      inputs: getInputs(namedInputs),
    },
    [options.submitTargetName]: {
      executor: `@nx/expo:submit`,
      dependsOn: [`^${options.submitTargetName}`],
      inputs: getInputs(namedInputs),
    },
  };

  return targets;
}

function getAppConfig(
  configFilePath: string,
  context: CreateNodesContext
): any {
  const resolvedPath = join(context.workspaceRoot, configFilePath);

  let module = load(resolvedPath);
  return module.default ?? module;
}

function getInputs(
  namedInputs: NxJsonConfiguration['namedInputs']
): TargetConfiguration['inputs'] {
  return [
    ...('production' in namedInputs
      ? ['default', '^production']
      : ['default', '^default']),
    {
      externalDependencies: ['react-native'],
    },
  ];
}

function getOutputs(projectRoot: string, dir: string) {
  if (projectRoot === '.') {
    return `{projectRoot}/${dir}`;
  } else {
    return `{workspaceRoot}/${projectRoot}/${dir}`;
  }
}

/**
 * Load the module after ensuring that the require cache is cleared.
 */
function load(path: string): any {
  // Clear cache if the path is in the cache
  if (require.cache[path]) {
    for (const k of Object.keys(require.cache)) {
      delete require.cache[k];
    }
  }

  // Then require
  return require(path);
}

function normalizeOptions(options: ExpoPluginOptions): ExpoPluginOptions {
  options ??= {};
  options.startTargetName ??= 'start';
  options.serveTargetName ??= 'serve';
  options.runIosTargetName ??= 'run-ios';
  options.runAndroidTargetName ??= 'run-android';
  options.exportTargetName ??= 'export';
  options.exportWebTargetName ??= 'export-web';
  options.prebuildTargetName ??= 'prebuild';
  options.installTargetName ??= 'install';
  options.buildTargetName ??= 'build';
  options.submitTargetName ??= 'submit';
  return options;
}
