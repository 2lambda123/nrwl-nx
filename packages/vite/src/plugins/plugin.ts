import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  detectPackageManager,
  joinPathFragments,
  readJsonFile,
  TargetConfiguration,
  workspaceRoot,
  writeJsonFile,
} from '@nx/devkit';
import { dirname, isAbsolute, join, relative, resolve } from 'path';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { loadConfigFromFile, UserConfig } from 'vite';
import { existsSync, readdirSync } from 'fs';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';
import { getLockFileName } from '@nx/js';

export interface VitePluginOptions {
  buildTargetName?: string;
  testTargetName?: string;
  serveTargetName?: string;
  previewTargetName?: string;
  serveStaticTargetName?: string;
}

const cachePath = join(projectGraphCacheDirectory, 'vite.hash');
const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};

const calculatedTargets: Record<
  string,
  Record<string, TargetConfiguration>
> = {};

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration>
> {
  return readJsonFile(cachePath);
}

function writeTargetsToCache(
  targets: Record<string, Record<string, TargetConfiguration>>
) {
  writeJsonFile(cachePath, targets);
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache(calculatedTargets);
  return [];
};

export const createNodes: CreateNodes<VitePluginOptions> = [
  '**/vite.config.{js,ts}',
  async (configFilePath, options, context) => {
    const projectRoot = dirname(configFilePath);
    // Do not create a project if package.json and project.json isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') &&
      !siblingFiles.includes('project.json')
    ) {
      return {};
    }

    options = normalizeOptions(options);

    const hash = calculateHashForCreateNodes(projectRoot, options, context, [
      getLockFileName(detectPackageManager(context.workspaceRoot)),
    ]);
    const targets = targetsCache[hash]
      ? targetsCache[hash]
      : await buildViteTargets(configFilePath, projectRoot, options, context);

    calculatedTargets[hash] = targets;

    return {
      projects: {
        [projectRoot]: {
          root: projectRoot,
          targets,
        },
      },
    };
  },
];

async function buildViteTargets(
  configFilePath: string,
  projectRoot: string,
  options: VitePluginOptions,
  context: CreateNodesContext
) {
  const viteConfig = await loadConfigFromFile(
    {
      command: 'build',
      mode: 'production',
    },
    configFilePath
  );

  const { buildOutputs, testOutputs } = getOutputs(
    projectRoot,
    viteConfig?.config
  );

  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration> = {};

  targets[options.buildTargetName] = await buildTarget(
    options.buildTargetName,
    namedInputs,
    buildOutputs,
    projectRoot
  );

  targets[options.serveTargetName] = serveTarget(projectRoot);

  targets[options.previewTargetName] = previewTarget(projectRoot);

  targets[options.testTargetName] = await testTarget(
    namedInputs,
    testOutputs,
    projectRoot
  );

  targets[options.serveStaticTargetName] = serveStaticTarget(options) as {};

  return targets;
}

async function buildTarget(
  buildTargetName: string,
  namedInputs: {
    [inputName: string]: any[];
  },
  outputs: string[],
  projectRoot: string
) {
  return {
    command: `vite build`,
    options: { cwd: joinPathFragments(projectRoot) },
    cache: true,
    dependsOn: [`^${buildTargetName}`],
    inputs: [
      ...('production' in namedInputs
        ? ['production', '^production']
        : ['default', '^default']),
      {
        externalDependencies: ['vite'],
      },
    ],
    outputs,
  };
}

function serveTarget(projectRoot: string) {
  const targetConfig: TargetConfiguration = {
    command: `vite serve`,
    options: {
      cwd: joinPathFragments(projectRoot),
    },
  };

  return targetConfig;
}

function previewTarget(projectRoot: string) {
  const targetConfig: TargetConfiguration = {
    command: `vite preview`,
    options: {
      cwd: joinPathFragments(projectRoot),
    },
  };

  return targetConfig;
}

async function testTarget(
  namedInputs: {
    [inputName: string]: any[];
  },
  outputs: string[],
  projectRoot: string
) {
  return {
    command: `vitest run`,
    options: { cwd: joinPathFragments(projectRoot) },
    cache: true,
    inputs: [
      ...('production' in namedInputs
        ? ['production', '^production']
        : ['default', '^default']),
      {
        externalDependencies: ['vitest'],
      },
    ],
    outputs,
  };
}

function serveStaticTarget(options: VitePluginOptions) {
  const targetConfig: TargetConfiguration = {
    executor: '@nx/web:file-server',
    options: {
      buildTarget: `${options.buildTargetName}`,
    },
  };

  return targetConfig;
}

function getOutputs(
  projectRoot: string,
  viteConfig: UserConfig
): {
  buildOutputs: string[];
  testOutputs: string[];
} {
  const { build, test } = viteConfig;
  const buildOutputs = ['{options.outputPath}'];
  const testOutputs = ['{options.reportsDirectory}'];

  function getOutput(path: string, projectRoot: string): string {
    if (path.startsWith('..')) {
      return join('{workspaceRoot}', join(projectRoot, path));
    } else if (isAbsolute(resolve(path))) {
      return `{workspaceRoot}/${relative(workspaceRoot, path)}`;
    } else {
      return join('{projectRoot}', path);
    }
  }

  if (build?.outDir) {
    buildOutputs.push(getOutput(build.outDir, projectRoot));
  }

  if (test?.coverage?.reportsDirectory) {
    testOutputs.push(getOutput(test.coverage.reportsDirectory, projectRoot));
  }

  return { buildOutputs, testOutputs };
}

function normalizeOptions(options: VitePluginOptions): VitePluginOptions {
  options ??= {};
  options.buildTargetName ??= 'build';
  options.serveTargetName ??= 'serve';
  options.previewTargetName ??= 'preview';
  options.testTargetName ??= 'test';
  options.serveStaticTargetName ??= 'serve-static';
  return options;
}
