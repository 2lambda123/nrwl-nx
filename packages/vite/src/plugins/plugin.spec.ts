import { CreateNodesContext } from '@nx/devkit';
import { createNodes } from './plugin';
import { TempFs } from 'nx/src/internal-testing-utils/temp-fs';

jest.mock('vite', () => ({
  loadConfigFromFile: jest.fn().mockImplementation(() => {
    return Promise.resolve({
      path: 'vite.config.ts',
      config: {},
      dependencies: [],
    });
  }),
}));

jest.mock('../utils/executor-utils', () => ({
  loadViteDynamicImport: jest.fn().mockResolvedValue({
    loadConfigFromFile: jest.fn().mockResolvedValue({
      path: 'vite.config.ts',
      config: {},
      dependencies: [],
    }),
  }),
}));

describe('@nx/vite/plugin', () => {
  let createNodesFunction = createNodes[1];
  let context: CreateNodesContext;
  describe('root project', () => {
    beforeEach(async () => {
      context = {
        nxJsonConfiguration: {
          // These defaults should be overridden by plugin
          targetDefaults: {
            build: {
              cache: false,
              inputs: ['foo', '^foo'],
            },
          },
          namedInputs: {
            default: ['{projectRoot}/**/*'],
            production: ['!{projectRoot}/**/*.spec.ts'],
          },
        },
        workspaceRoot: '',
      };
    });

    afterEach(() => {
      jest.resetModules();
    });

    it('should create nodes', async () => {
      const nodes = await createNodesFunction(
        'vite.config.ts',
        {
          buildTargetName: 'build',
          serveTargetName: 'serve',
          previewTargetName: 'preview',
          testTargetName: 'test',
          serveStaticTargetName: 'serve-static',
        },
        context
      );

      expect(nodes).toMatchSnapshot();
    });
  });

  // some issue wiht the tempfs
  describe('not root project', () => {
    const tempFs = new TempFs('test');
    beforeEach(() => {
      context = {
        nxJsonConfiguration: {
          namedInputs: {
            default: ['{projectRoot}/**/*'],
            production: ['!{projectRoot}/**/*.spec.ts'],
          },
        },
        workspaceRoot: tempFs.tempDir,
      };

      tempFs.createFileSync(
        'my-app/project.json',
        JSON.stringify({ name: 'my-app' })
      );
      tempFs.createFileSync('my-app/vite.config.ts', '');
    });

    afterEach(() => {
      jest.resetModules();
    });

    it('should create nodes', async () => {
      const nodes = await createNodesFunction(
        'my-app/vite.config.ts',
        {
          buildTargetName: 'build-something',
          serveTargetName: 'my-serve',
          previewTargetName: 'preview-site',
          testTargetName: 'vitest',
          serveStaticTargetName: 'serve-static',
        },
        context
      );

      expect(nodes).toMatchSnapshot();
    });
  });
});
