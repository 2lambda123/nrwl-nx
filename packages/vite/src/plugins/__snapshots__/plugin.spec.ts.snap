// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nx/vite/plugin not root project should create nodes 1`] = `
{
  "projects": {
    "my-app": {
      "root": "my-app",
      "targets": {
        "build-something": {
          "cache": true,
          "command": "pnpm exec vite build",
          "inputs": [
            "production",
            "^production",
          ],
          "options": {
            "cwd": "my-app",
          },
          "outputs": [
            "{options.outputPath}",
          ],
        },
        "my-serve": {
          "command": "pnpm exec vite serve",
          "options": {
            "cwd": "my-app",
          },
        },
        "preview-site": {
          "command": "pnpm exec vite preview",
          "options": {
            "cwd": "my-app",
          },
        },
        "serve-static": {
          "executor": "@nx/web:file-server",
          "options": {
            "buildTarget": "build-something",
          },
        },
        "vitest": {
          "cache": true,
          "command": "pnpm exec vitest run",
          "inputs": [
            "default",
            "^production",
          ],
          "options": {
            "cwd": "my-app",
          },
          "outputs": [
            "{options.reportsDirectory}",
          ],
        },
      },
    },
  },
}
`;

exports[`@nx/vite/plugin root project should create nodes 1`] = `
{
  "projects": {
    ".": {
      "root": ".",
      "targets": {
        "build": {
          "cache": true,
          "command": "pnpm exec vite build",
          "inputs": [
            "production",
            "^production",
          ],
          "options": {
            "cwd": ".",
          },
          "outputs": [
            "{options.outputPath}",
          ],
        },
        "preview": {
          "command": "pnpm exec vite preview",
          "options": {
            "cwd": ".",
          },
        },
        "serve": {
          "command": "pnpm exec vite serve",
          "options": {
            "cwd": ".",
          },
        },
        "serve-static": {
          "executor": "@nx/web:file-server",
          "options": {
            "buildTarget": "build",
          },
        },
        "test": {
          "cache": true,
          "command": "pnpm exec vitest run",
          "inputs": [
            "default",
            "^production",
          ],
          "options": {
            "cwd": ".",
          },
          "outputs": [
            "{options.reportsDirectory}",
          ],
        },
      },
    },
  },
}
`;
